File: engines/cpp/levenshtein_test.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include "distance.h"

static int passed = 0;

void check(const std::string& name, int got, int want) {
    if (got != want) {
        std::cerr << "FAIL " << name << ": got " << got << ", want " << want << "\n";
        std::exit(1);
    }
    std::cout << "PASS " << name << "\n";
    passed++;
}

int main() {
    check("identical",          distance("cat", "cat"),       0);
    check("case_insensitive",   distance("cat", "Cat"),       0);
    check("case_insensitive_2", distance("CAT", "cat"),       0);
    check("substitution",       distance("cat", "bat"),       1);
    check("deletion",           distance("cat", "ca"),        1);
    check("insertion",          distance("cat", "cart"),      1);
    check("empty_a",            distance("", "abc"),          3);
    check("empty_b",            distance("abc", ""),          3);
    check("both_empty",         distance("", ""),             0);
    check("kitten_sitting",     distance("kitten", "sitting"),3);
    check("saturday_sunday",    distance("saturday", "sunday"),3);

    std::cout << "\n" << passed << " tests passed.\n";
    return 0;
}

File: engines/go/levenshtein_test.go
package main

import (
	"testing"
)

func TestDistance(t *testing.T) {
	cases := []struct {
		a, b string
		want int
	}{
		{"cat", "cat", 0},
		{"cat", "Cat", 0},  // case-insensitive
		{"cat", "bat", 1},  // substitution
		{"cat", "ca", 1},   // deletion
		{"cat", "cart", 1}, // insertion
		{"", "abc", 3},
		{"abc", "", 3},
		{"", "", 0},
		{"kitten", "sitting", 3}, // classic example
		{"saturday", "sunday", 3},
	}

	for _, c := range cases {
		got := distance(c.a, c.b)
		if got != c.want {
			t.Errorf("distance(%q, %q) = %d, want %d", c.a, c.b, got, c.want)
		}
	}
}

func TestSearch(t *testing.T) {
	dict := []string{"bat", "hat", "sat", "cat", "car", "bar", "dog", "log", "fog"}

	t.Run("exact delta=1 from cat", func(t *testing.T) {
		res := Search("cat", 1, 10, dict)
		want := map[string]bool{"bat": true, "hat": true, "sat": true, "car": true}
		if len(res.Words) != len(want) {
			t.Fatalf("got %v, want %v", res.Words, want)
		}
		for _, w := range res.Words {
			if !want[w] {
				t.Errorf("unexpected result %q", w)
			}
		}
	})

	t.Run("limit is respected", func(t *testing.T) {
		res := Search("cat", 1, 2, dict)
		if len(res.Words) != 2 {
			t.Errorf("got %d results, want 2", len(res.Words))
		}
	})

	t.Run("no results", func(t *testing.T) {
		res := Search("zzz", 1, 10, dict)
		if len(res.Words) != 0 {
			t.Errorf("expected no results, got %v", res.Words)
		}
	})

	t.Run("delta=0 returns only exact match", func(t *testing.T) {
		res := Search("cat", 0, 10, dict)
		if len(res.Words) != 1 || res.Words[0] != "cat" {
			t.Errorf("got %v, want [cat]", res.Words)
		}
	})

	t.Run("case-insensitive match", func(t *testing.T) {
		mixedDict := []string{"Bat", "HAT", "sat"}
		res := Search("cat", 1, 10, mixedDict)
		if len(res.Words) != 3 {
			t.Errorf("got %v, want [Bat HAT sat]", res.Words)
		}
	})
}
